### PR TITLE
Added FUSE stub program implementation

### PR DESCRIPTION
**Pull request by tropuq**
See #2

### CODE REVIEW

File: fs/fuse_apps/fuse_block_device.c [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+	if (ret == -1)
+		return -errno;
+	fd = ret;
+
+	ret = ioctl(fd, BLKGETSIZE64, &res);
+	if (ret == -1)
+		res = -errno;
+
+	close(fd);
+
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c:56]: This newline is not needed

File: fs/fuse_apps/fuse_block_device.c [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+static const struct fuse_opt option_spec[] = {
+	OPTION("--blockdev=%s", blockdev),
+	OPTION("--filename=%s", filename),
+	OPTION("-h", show_help),
+	OPTION("--help", show_help),
+	FUSE_OPT_END
+};
+
+static ssize_t get_blockdev_size(const char *path)
+{
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]: { should be in the same line as function definition AFAIK

File: fs/fuse_apps/fuse_block_device.c [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stddef.h>
+#include <assert.h>
+#include <sys/ioctl.h>
+#include <linux/fs.h>
+#include <unistd.h>
+#include <sys/stat.h>
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]: Sort includes

File: fs/fuse_apps/fuse_block_device.c [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+		res = -errno;
+
+	close(fd);
+
+	return res;
+}
+
+static int block_getattr(const char *path, struct stat *stbuf,
+			 struct fuse_file_info *fi)
+{
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]: { should be in the same line as function definition AFAIK

File: fs/fuse_apps/fuse_block_device.c [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+/*
+ * Command line options
+ *
+ * We can't set default values for the char* fields here because
+ * fuse_opt_parse would attempt to free() them when the user specifies
+ * different values on the command line.
+ */
+static struct options {
+	const char *blockdev;
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]: Convert tabs to spaces everywhere

File: fs/fuse_apps/fuse_block_device.c [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+static int block_readdir(const char *path, void *buf, fuse_fill_dir_t filler,
+			 off_t offset, struct fuse_file_info *fi,
+			 enum fuse_readdir_flags flags)
+{
+	(void) offset;
+	(void) fi;
+	(void) flags;
+
+	if (strcmp(path, "/") != 0)
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]: Add braces around this

File: fs/fuse_apps/fuse_block_device.c [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+	filler(buf, ".", NULL, 0, 0);
+	filler(buf, "..", NULL, 0, 0);
+	filler(buf, options.filename, NULL, 0, 0);
+
+	return 0;
+}
+
+static int block_open(const char *path, struct fuse_file_info *fi)
+{
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]: { should be in the same line as function definition AFAIK

File: fs/fuse_apps/fuse_block_device.c [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+	fd = open(options.blockdev, fi->flags);
+	if (fd == -1)
+		return -errno;
+
+	fi->fh = fd;
+	return 0;
+}
+
+static int block_release(const char *path, struct fuse_file_info *fi)
+{
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]: { should be in the same line as function definition AFAIK

File: fs/fuse_apps/fuse_block_device.c [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+	(void) path;
+	close(fi->fh);
+	return 0;
+}
+
+static int block_read(const char *path, char *buf, size_t size, off_t offset,
+		      struct fuse_file_info *fi)
+{
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]: { should be in the same line as function definition AFAIK

File: fs/fuse_apps/fuse_block_device.c [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+	int fd, res;
+
+	if (strcmp(path+1, options.filename) != 0)
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]: add braces

File: fs/fuse_apps/fuse_block_device.c [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+		return -ENOENT;
+
+	if (fi == NULL)
+		fd = open(path, O_RDONLY);
+	else
+		fd = fi->fh;
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]: add braces

File: fs/fuse_apps/fuse_block_device.c [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+
+	if (fd == -1)
+		return -errno;
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]: add braces here and below

File: fs/fuse_apps/fuse_block_device.c [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+
+	/* Set defaults -- we have to use strdup so that
+	   fuse_opt_parse can free the defaults if other
+	   values are specified */
+	options.blockdev = strdup("/dev/nvme0n1");
+	options.filename = strdup("blockdev");
+
+	/* Parse options */
+	if (fuse_opt_parse(&args, &options, option_spec, NULL) == -1)
+		return 1;
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]: add braces

File: fs/fuse_apps/fuse_block_device.c [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+	       "\n");
+}
+
+int main(int argc, char *argv[])
+{
+	int ret;
+	struct fuse_args args = FUSE_ARGS_INIT(argc, argv);
+
+	/* Set defaults -- we have to use strdup so that
+	   fuse_opt_parse can free the defaults if other
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]:
```suggestion
	   fuse_opt_parse() can free the defaults if other
```

File: fs/fuse_apps/fuse_block_device.c [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+#include <sys/ioctl.h>
+#include <linux/fs.h>
+#include <unistd.h>
+#include <sys/stat.h>
+
+/*
+ * Command line options
+ *
+ * We can't set default values for the char* fields here because
+ * fuse_opt_parse would attempt to free() them when the user specifies
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]:
```suggestion
 * fuse_opt_parse() would attempt to free() them when the user specifies
```

File: fs/fuse_apps/fuse_block_device.c [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+	   values are specified */
+	options.blockdev = strdup("/dev/nvme0n1");
+	options.filename = strdup("blockdev");
+
+	/* Parse options */
+	if (fuse_opt_parse(&args, &options, option_spec, NULL) == -1)
+		return 1;
+
+	/* When --help is specified, first print our own file-system
+	   specific help text, then signal fuse_main to show
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]:
```suggestion
	   specific help text, then signal fuse_main() to show
```

**Review by varqox — CHANGES_REQUESTED** [review]

File: fs/fuse_apps/fuse_block_device.c [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+    filler(buf, "..", NULL, 0, 0);
+    filler(buf, options.filename, NULL, 0, 0);
+
+    return 0;
+}
+
+static int block_open(const char *path, struct fuse_file_info *fi) {
+    int fd;
+
+    if (strcmp(path+1, options.filename) != 0) {
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]:
```suggestion
    if (strcmp(path + 1, options.filename) != 0) {
```

File: fs/fuse_apps/fuse_block_device.c [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+    (void) path;
+    close(fi->fh);
+    return 0;
+}
+
+static int block_read(const char *path, char *buf, size_t size, off_t offset,
+        struct fuse_file_info *fi) {
+    int fd, res;
+
+    if (strcmp(path+1, options.filename) != 0) {
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]:
```suggestion
    if (strcmp(path + 1, options.filename) != 0) {
```

File: fs/fuse_apps/fuse_block_device.c [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+    /* Parse options */
+    if (fuse_opt_parse(&args, &options, option_spec, NULL) == -1) {
+        return 1;
+    }
+
+    /* When --help is specified, first print our own file-system
+       specific help text, then signal fuse_main() to show
+       additional help (by adding `--help` to the options again)
+       without usage: line (by setting argv[0] to the empty
+       string) */
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]: I totally don't understand this part of the comment.

File: fs/fuse_apps/fuse_block_device.c [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+        struct fuse_file_info *fi) {
+    (void) fi;
+    int res = 0;
+    ssize_t blockdev_size;
+
+    memset(stbuf, 0, sizeof(struct stat));
+    if (strcmp(path, "/") == 0) {
+        stbuf->st_mode = S_IFDIR | 0755;
+        stbuf->st_nlink = 2;
+    } else if (strcmp(path+1, options.filename) == 0) {
```
Comment (varqox) [inline fs/fuse_apps/fuse_block_device.c]:
```suggestion
    } else if (strcmp(path + 1, options.filename) == 0) {
```

**Review by varqox — CHANGES_REQUESTED** [review]

File: fs/fuse_apps/fuse_block_device.c (hunk 19 above)
Comment (tropuq) [reply fs/fuse_apps/fuse_block_device.c]: It means it won't execute `show_help()` again. _usage: line_  refers to `printf("usage: ...` part (line 169)

Comment (tropuq) [reply fs/fuse_apps/fuse_block_device.c]: I rewrote that part. I think it's better now.

**Review by varqox — APPROVED** [review]

**Pull request closed without merging**